contracts/examples/bonding-curve/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+#![no_std]
+#![allow(unused_attributes)]
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+mod bc_function;
+use bc_function::{BCFunction, CurveArguments, SupplyType, Token};
+
+mod linear_function;
+use linear_function::LinearFunction;
+
+const TOKEN_NUM_DECIMALS: usize = 18;
+
+#[elrond_wasm_derive::contract]
+pub trait BondingCurve {
+	#[init]
+	fn init(&self, issued_token: Token, exchanging_token: Token) {
+		self.issued_token().set(&issued_token);
+		self.exchanging_token().set(&exchanging_token);
+	}
+
+	// endpoint - owner-only
+
+	#[payable("*")]
+	#[endpoint(issueToken)]
+	fn issue_token(
+		&self,
+		initial_supply: Self::BigUint,
+		#[payment] issue_cost: Self::BigUint,
+	) -> SCResult<AsyncCall<Self::SendApi>> {
+		only_owner!(self, "only owner may call this function");
+
+		require!(
+			self.issued_token().is_empty(),
+			"wrapped egld was already issued"
+		);
+
+		let caller = self.blockchain().get_caller();
+
+		self.issue_started_event(
+			&caller,
+			&self.issued_token().get().identifier,
+			&initial_supply,
+		);
+
+		Ok(ESDTSystemSmartContractProxy::new_proxy_obj(self.send())
+			.issue_fungible(
+				issue_cost,
+				&self.issued_token().get().name,
+				&self.issued_token().get().identifier.into_boxed_bytes(),
+				&initial_supply,
+				FungibleTokenProperties {
+					num_decimals: TOKEN_NUM_DECIMALS,
+					can_freeze: false,
+					can_wipe: false,
+					can_pause: false,
+					can_mint: true,
+					can_burn: false,
+					can_change_owner: true,
+					can_upgrade: true,
+					can_add_special_roles: false,
+				},
+			)
+			.async_call()
+			.with_callback(self.callbacks().issue_callback(&caller)))
+	}
+
+	#[callback]
+	fn issue_callback(
+		&self,
+		caller: &Address,
+		#[payment] amount: Self::BigUint,
+		#[call_result] result: AsyncCallResult<()>,
+	) {
+		let token = self.issued_token().get().identifier;
+		match result {
+			AsyncCallResult::Ok(()) => {
+				self.issue_success_event(caller, &token, &amount);
+				self.supply().set(&amount);
+				self.balance().set(&amount);
+			},
+			AsyncCallResult::Err(message) => {
+				self.issue_failure_event(caller, message.err_msg.as_slice());
+				if token.is_egld() && amount > 0 {
+					self.send().direct_egld(caller, &amount, &[]);
+				}
+			},
+		}
+	}
+
+	#[endpoint(mintToken)]
+	fn mint_token(&self, amount: Self::BigUint) -> SCResult<AsyncCall<Self::SendApi>> {
+		only_owner!(self, "only owner may call this function");
+
+		let token = self.issued_token().get().identifier;
+		if self.issued_token().is_empty() {
+			return Err(SCError::from(BoxedBytes::from_concat(&[
```

Comment:
Keep in mind this brings a lot of bloat. Also, if self.issued_token() is empty, you have no token name to provide for the error message.

Since you have only one "issued_token", there is no reason to provide its name. Simply
return sc_error!("Must issue token before minting");